### PR TITLE
Hive

### PR DESCRIPTION
**Pull request by mrocklin**
Some updates for Hive

1.  Clean up ssh-hive tests to be less stateful
2.  Handle CSV dialects from directories
3.  Reduce the amount of full-database reflection that happens (this is expensive)
4.  Accept stored_as= and external= kwargs to create non-textfile hive tables

TODO: insert syntax in PyHive doesn't work

### CODE REVIEW

File: odo/backends/hdfs.py [hunk 1]
```diff
@@ -252,17 +281,24 @@ def create_hive_statement(tbl_name, dshape, path=None, table_type='',
     # Column names and types from datashape
     ds = dshape or discover(data)
```
Comment (cpcloud) [inline odo/backends/hdfs.py]: it looks like there's an undefined variable `data` here

File: odo/backends/sql.py [hunk 2]
```diff
@@ -309,6 +309,18 @@ def append_anything_to_sql_Table(t, c, **kwargs):
 def append_anything_to_sql_Table(t, o, **kwargs):
     return append(t, convert(Iterator, o, **kwargs), **kwargs)
 
+@append.register(sa.Table, sa.Table)
```
Comment (cpcloud) [inline odo/backends/sql.py]: should there be a test that this fails when trying to insert, say, a sqlite table into a hive table?

Comment (cpcloud) [inline odo/backends/sql.py]: actually, it looks like that might actually *pass*

File: docs/source/hive.rst [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    Hive -> Hive
+
+And as with all SQL systems through SQLAlchemy we can convert a Hive table to a
+Python Iterator, though this is somewhat slow::
+
+    Hive -> Iterator
+
+
+Impala
```
Comment (cpcloud) [inline docs/source/hive.rst]: I wonder if a small list of technologies that odo handles that work with hive might be nice.

Comment (mrocklin) [inline docs/source/hive.rst]: I'm not familiar enough with spark+hive+odo interactions.  Do we have this?

File: odo/backends/sql.py (hunk 2 above)
Comment (mrocklin) [inline odo/backends/sql.py]: This check occurs just below in the append Select to Table call

File: odo/backends/hdfs.py (hunk 1 above)
Comment (mrocklin) [inline odo/backends/hdfs.py]: Fixed

**Comment by mrocklin** [comment]
Other comments?  I'd like to merge this for the release.

**Pull request merged**